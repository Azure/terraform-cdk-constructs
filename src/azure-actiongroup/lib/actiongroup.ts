import { MonitorActionGroup } from "@cdktf/provider-azurerm/lib/monitor-action-group";
import { ResourceGroup } from "@cdktf/provider-azurerm/lib/resource-group";
import * as cdktf from "cdktf";
import { Construct } from "constructs";
import { AzureResource } from "../../core-azure/lib";
import * as model from "../model";

export class ActionGroup extends AzureResource {
  readonly props: model.ActionGroupProps;
  public resourceGroup: ResourceGroup;
  public id: string;
  /**
   * Manages an Azure Monitor Action Group, which is used to trigger actions or notifications based on alerts or conditions met within Azure Monitor.
   *
   * An Action Group in Azure Monitor defines a collection of individual actions that are triggered when the conditions of an associated alert rule are met. Actions can include sending emails, triggering Azure Functions, calling webhooks, and more. This class allows for configuring and managing these actions, making it essential for setting up comprehensive monitoring and response systems in Azure applications.
   *
   * @param scope - The scope in which to define this construct, typically representing the Cloud Development Kit (CDK) stack.
   * @param id - The unique identifier for this instance of the Action Group.
   * @param props - Configuration properties for the Action Group. These properties may include:
   *                - `name`: The name of the Action Group.
   *                - `resourceGroup`: Optional. Reference to the resource group for deployment.
   *                - `shortName`: A shorter name for the Action Group used in notifications.
   *                - `enabled`: Specifies if the Action Group is active. Defaults to true.
   *                - `location`: The Azure region where the Action Group is hosted. Defaults to global.
   *                - `tags`: A dictionary of tags to apply to the Action Group for resource management.
   *                - Receivers: Configurations for various types of notifications (e.g., email, SMS, webhook).
   *
   * Example usage:
   * ```typescript
   * const actionGroup = new ActionGroup(this, 'myActionGroup', {
   *   name: 'criticalAlertsGroup',
   *   resourceGroup: myResourceGroup,
   *   shortName: 'Alerts',
   *   location: 'East US',
   *   emailReceivers: [{
   *     name: 'admin',
   *     emailAddress: 'admin@example.com'
   *   }],
   *   smsReceivers: [{
   *     name: 'adminSms',
   *     countryCode: '1',
   *     phoneNumber: '5551234567'
   *   }],
   *   tags: {
   *     environment: 'production'
   *   }
   * });
   * ```
   * This setup creates an Action Group that sends email and SMS notifications when triggered by an alert.
   */
  constructor(scope: Construct, id: string, props: model.ActionGroupProps) {
    super(scope, id);

    this.props = props;
    this.resourceGroup = this.setupResourceGroup(props);

    /**
     * Define default values.
     */
    const defaults = {
      enabled: props.enabled ?? true,
      location: props.location ?? "global",
      tags: props.tags ?? {},
    };

    // Create the Action Group with the provided properties
    const azurermMonitorActionGroup = new MonitorActionGroup(
      this,
      "actiongroup",
      {
        ...defaults,
        name: props.name,
        resourceGroupName: this.resourceGroup.name,
        shortName: props.shortName,
        armRoleReceiver: cdktf.listMapper(
          model.monitorActionGroupArmRoleReceiverToTerraform,
          true,
        )(this.props.armRoleReceivers),

        emailReceiver: cdktf.listMapper(
          model.monitorActionGroupEmailReceiverToTerraform,
          true,
        )(this.props.emailReceivers),

        voiceReceiver: cdktf.listMapper(
          model.monitorActionGroupVoiceReceiverToTerraform,
          true,
        )(this.props.voiceReceivers),

        smsReceiver: cdktf.listMapper(
          model.monitorActionGroupSmsReceiverToTerraform,
          true,
        )(this.props.smsReceivers),

        webhookReceiver: cdktf.listMapper(
          model.monitorActionGroupWebhookReceiverToTerraform,
          true,
        )(this.props.webhookReceivers),

        eventHubReceiver: cdktf.listMapper(
          model.monitorActionGroupEventHubReceiverToTerraform,
          true,
        )(this.props.eventHubReceivers),

        azureAppPushReceiver: cdktf.listMapper(
          model.monitorActionGroupAzureAppPushReceiverToTerraform,
          true,
        )(this.props.azureAppPushReceivers),

        logicAppReceiver: cdktf.listMapper(
          model.monitorActionGroupLogicAppReceiverToTerraform,
          true,
        )(this.props.logicAppReceivers),
      },
    );

    // Terraform Outputs
    this.id = azurermMonitorActionGroup.id;

    const cdktfTerraformOutputActionGroupId = new cdktf.TerraformOutput(
      this,
      "id",
      {
        value: azurermMonitorActionGroup.id,
      },
    );

    /*This allows the Terraform resource name to match the original name. You can remove the call if you don't need them to match.*/
    cdktfTerraformOutputActionGroupId.overrideLogicalId("id");
  }
}
