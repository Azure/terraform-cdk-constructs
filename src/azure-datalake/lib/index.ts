export * from "./filesystem";
export * from "./path";
export * from "./datalake";
