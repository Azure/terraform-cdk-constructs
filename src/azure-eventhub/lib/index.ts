export * from "./authorization";
export * from "./eventhubcluster";
export * from "./consumer";
export * from "./instance";
export * from "./kusto-connection";
export * from "./namespace";
