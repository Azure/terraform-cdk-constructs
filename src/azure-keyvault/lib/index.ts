export * from './certificate'
export * from './key'
export * from './policy'
export * from './secret'
export * from './vault'

