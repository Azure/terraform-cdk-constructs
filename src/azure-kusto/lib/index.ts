export * from "./cluster";
export * from "./compute-specification";
export * from "./database";
