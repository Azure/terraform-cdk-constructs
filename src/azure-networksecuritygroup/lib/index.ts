export * from './preconfigured-rules';
export * from './securitygroup';
