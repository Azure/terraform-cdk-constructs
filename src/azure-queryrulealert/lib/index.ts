export * from './query-rule-alert'