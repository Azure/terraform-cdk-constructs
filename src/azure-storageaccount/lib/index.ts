export * from './account';
export * from './container';
export * from './fileshare';
export * from './queue';
export * from './table';
