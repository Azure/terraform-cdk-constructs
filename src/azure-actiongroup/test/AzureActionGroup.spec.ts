import { Testing, TerraformStack } from 'cdktf';
import { exampleAzureActionGroup } from './ExampleAzureActionGroup'
import 'cdktf/lib/testing/adapters/jest';
import { ActionGroup } from "..";
import { AzurermProvider } from "@cdktf/provider-azurerm/lib/provider";
import * as util from "../../util/azureTenantIdHelpers";


describe('Action Group With Defaults', () => {
  let stack: TerraformStack;
  let fullSynthResult: any;

  beforeEach(() => {
    const app = Testing.app();
    stack = new TerraformStack(app, "test");

    new AzurermProvider(stack, "azureFeature", { features: {} });
    new ActionGroup(stack, 'testAzureActionGroupDefaults', {
      name: "testactiongroup",
      location: 'eastus',
      resourceGroupName: "rg-test",
      shortName: "testshortn",
    });

    fullSynthResult = Testing.fullSynth(stack); // Save the result for reuse
  });

  it("renders an Action Group with defaults and checks snapshot", () => {
    expect(
      Testing.synth(stack)
    ).toMatchSnapshot(); // Compare the already prepared stack
  });

  it("check if the produced terraform configuration is valid", () => {
    expect(fullSynthResult).toBeValidTerraform(); // Use the saved result
  });

  it("check if this can be planned", () => {
    expect(fullSynthResult).toPlanSuccessfully(); // Use the saved result
  });
});


describe('Action Group Example', () => {

  it("renders and checks snapshot", () => {
    // Need to remove the tenant_id from the snapshot as it will change wherever the test is run
    const output = Testing.synth(new exampleAzureActionGroup(Testing.app(), "testAzureActionGroup"));
    const myObject: Record<string, any> = JSON.parse(output);

    expect(util.removeTenantIdFromSnapshot(myObject)).toMatchSnapshot();
  });

  it("check if the produced terraform configuration is valid", () => {
    // We need to do a full synth to plan the terraform configuration
    expect(Testing.fullSynth(new exampleAzureActionGroup(Testing.app(), "testAzureActionGroup"))).toBeValidTerraform();
  });

  it("check if this can be planned", () => {

    // We need to do a full synth to plan the terraform configuration
    expect(Testing.fullSynth(new exampleAzureActionGroup(Testing.app(), "testAzureActionGroup"))).toPlanSuccessfully();
  });
});