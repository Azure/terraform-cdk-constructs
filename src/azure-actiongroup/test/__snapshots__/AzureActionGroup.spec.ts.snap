// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Action Group Example renders and checks snapshot 1`] = `
{
  "output": {
    "id": {
      "value": "\${azurerm_monitor_action_group.testAzureActionGroup_actiongroup_73C559FB.id}",
    },
  },
  "provider": {
    "azurerm": [
      {
        "features": {
          "resource_group": {
            "prevent_deletion_if_contains_resources": false,
          },
        },
      },
    ],
  },
  "resource": {
    "azurerm_monitor_action_group": {
      "testAzureActionGroup_actiongroup_73C559FB": {
        "dynamic": {
          "email_receiver": {
            "content": {
              "email_address": "\${email_receiver.value.emailAddress}",
              "name": "\${email_receiver.value.name}",
              "use_common_alert_schema": "\${email_receiver.value.useCommonAlertSchema}",
            },
            "for_each": [
              {
                "emailAddress": "test1@email.com",
                "name": "testemail1",
                "useCommonAlertSchema": false,
              },
              {
                "emailAddress": "test2@email.com",
                "name": "testemail2",
                "useCommonAlertSchema": false,
              },
            ],
          },
        },
        "enabled": true,
        "location": "global",
        "name": "testactiongroup",
        "resource_group_name": "\${azurerm_resource_group.rg.name}",
        "short_name": "testshortn",
        "tags": {},
      },
    },
    "azurerm_resource_group": {
      "rg": {
        "location": "eastus",
        "name": "rg-\${var.name}",
      },
    },
  },
  "terraform": {
    "required_providers": {
      "azurerm": {
        "source": "azurerm",
        "version": "3.70.0",
      },
    },
  },
  "variable": {
    "name": {
      "default": "test",
      "description": "System name used to randomize the resources",
      "type": "string",
    },
  },
}
`;

exports[`Action Group With Defaults renders an Action Group with defaults and checks snapshot 1`] = `
"{
  "output": {
    "id": {
      "value": "\${azurerm_monitor_action_group.testAzureActionGroupDefaults_actiongroup_ADD914FD.id}"
    }
  },
  "provider": {
    "azurerm": [
      {
        "features": {
        }
      }
    ]
  },
  "resource": {
    "azurerm_monitor_action_group": {
      "testAzureActionGroupDefaults_actiongroup_ADD914FD": {
        "enabled": true,
        "location": "eastus",
        "name": "testactiongroup",
        "resource_group_name": "rg-test",
        "short_name": "testshortn",
        "tags": {
        }
      }
    }
  },
  "terraform": {
    "required_providers": {
      "azurerm": {
        "source": "azurerm",
        "version": "3.70.0"
      }
    }
  }
}"
`;
